src/Builder/FrontendDialectHelper.cpp

```diff
@@ -96,20 +96,22 @@ struct TransformValueToONNXData<int64_t> {
 
 // Helper method for constructing an array attribute from a model input.
 template <typename T>
-static T* CreateArrayAttribute(onnx::TensorProto initializer, int *size) {
+static std::vector<T> CreateArrayAttribute(onnx::TensorProto initializer) {
```

Comment:
Now that we actually returns the vector, we may want to return `const std::vector<T>&`?

Reply:
Tian, `const std::vector<T>&` is still refer to the local data of the function.

Reply:
If I remember correctly, this is a special case - the lifetime of const reference to a temporary object will be automatically extended. Essentially the underlying vector gets destroyed only when the reference to it goes out of scope.

Reply:
Let me provide a reference for completeness - https://en.cppreference.com/w/cpp/language/lifetime

> The lifetime of a temporary object may be extended by binding to a const lvalue reference or to an rvalue reference (since C++11), see reference initialization for details.

Reply:
I changed `static std::vector<T> CreateArrayAttribute` to `static const std::vector<T> &CreateArrayAttribute` but it did not work. Anything else do I need to modify?

In this section about lifetime of a temprary, https://en.cppreference.com/w/cpp/language/reference_initialization#Lifetime_of_a_temporary, I see this exception:
>a temporary bound to a return value of a function in a return statement is not extended: it is destroyed immediately at the end of the return expression. Such function always returns a dangling reference.

Reply:
Aha, sorry for my unhelpful comments, I've played around with it a bit to see what's the right way to avoid this potentially expensive copy using const reference and arrived at the following code diff -

```diff
case (onnx::TensorProto::FLOAT): {
-      float *typeArray =
-          CreateArrayAttribute<float>(initializer, &length);
-      std::vector<float> arrayAttrInitializer(
-       typeArray, typeArray + length);
-      llvm::ArrayRef<float> array(typeArray, length);
+      const auto& arrayAttrInitializer =
+          CreateArrayAttribute<float>(initializer);
       elementType = builder.getF32Type();
       tensorType = mlir::RankedTensorType::get(tensorDims, elementType);
-      constantDenseAttribute = mlir::DenseElementsAttr::get(tensorType, array);
+      constantDenseAttribute = mlir::DenseElementsAttr::get(
+          tensorType, llvm::makeArrayRef(arrayAttrInitializer));
       break;
     }
```

It seems to compile and the tests work fine.

Reply:
Sorry for the confusion, I think there is no problem with our tests since they do not have initialized tensors (e.g. learned weights). Also, I did not see problems with a small MNIST model but large models like ResNet50. As I mentioned in this patch's description, many weights were successfully imported until a point where maybe there were not enough memory for temporary references.

Reply:
@tjingrant Using `const auto&` works well. Thanks!

Reply:
Thanks! I also checked with ResNet50 locally just now and the initializer importing logic succeeds.